### PR TITLE
Refuse to read a negative number of bytes; test this as well.

### PR DESCRIPTION
**Pull request by yomimono**
Also, refer to a more generic name than FS_unix in test code.

Signed-off-by: Mindy Preston <mindy.preston@docker.com>

### CODE REVIEW

**Comment by yomimono** [comment]
The discovery of this is related to https://github.com/mirage/ocaml-cstruct/issues/86 , but I think the refusal is reasonable to implement here as well.

File: lib/fs_common.ml [hunk 1]
```diff
@@ -57,27 +57,31 @@ let map_error err reqd_string =
...
-        | len ->
-           return (Some (Cstruct.sub buf 0 (min len reqd_len)))
-      )
-    in
-    Lwt_stream.to_list st >>= fun bufs ->
-    match reqd_len with
-    | 0 -> return (`Ok [])
-    | n -> return (`Ok bufs)
-  with Unix.Unix_error (ex, _, _) -> return (map_error ex name)
+    try_lwt
```
Comment (samoht) [inline lib/fs_common.ml]: while you are it, can you remove camlp4 syntax extension? `Lwt.catch` works well in that case :-)

File: lib/fs_common.ml [hunk 2]
```diff
@@ -57,27 +57,31 @@ let map_error err reqd_string =
                                     (reqd_string ^ ": error " ^ string_of_int i))
 
 let read_impl base name off reqd_len =
+  if reqd_len < 0 then return (`Error (`Unknown_error "can't read negative bytes"
+                                      ))
```
Comment (samoht) [inline lib/fs_common.ml]: this parenthesis look lonely

**Comment by yomimono** [comment]
@samoht , think this is mergeable?

**Comment by samoht** [comment]
Ideally you could also remove `lwt.syntax` for the _oasis but I guess this could be done in a separate PR.

**Comment by hannesm** [comment]
I'm actually not sure about invalid input, and which module should validate that?  Certainly, crashes are not nice, but shouldn't there (isn't there) a library which consumes a BLOCK device and presents a BLOCK device, just inserting bounds checks?

In the end, a negative read (or other invalid reads/writes/...) should also be prevented from mirage-fs-xen, mirage-block-ramdisk, mirage-block-ccm, ... ?  For me, a BLOCK in itself should not perform any bounds checks, but pretend that the input is already validated.  (Of course, this can also be another thread of discussion, and a separate PR.  Overall this PR improves things!)

**Comment by yomimono** [comment]
@samoht - I'm fine to pull all the `lwt.syntax` stuff out of this repo but I'd prefer to do it in a separate PR.

@hannesm - I share your lack of certainty about where exactly is generally appropriate for bounds checking in the stack, but I think enforcing that offsets can't be negative makes sense for the FS layer specifically (including mirage-fs-xen).  If we're talking about root causes for this bug, I think the Cstruct.uint32 != int32 issue linked above is more sensible to target.

**Comment by djs55** [comment]
I started creating a BLOCK device wrapper here:

https://github.com/mirage/mirage-block/blob/95f82605a92571843b6aba7c1408dd79d6a4b534/lib/mirage_block.mli#L87

Currently it just checks lengths of the supplied buffers (must be multiples
of sector_size) but it could be extended with bounds checks too:

https://github.com/mirage/mirage-block/blob/master/lib/mirage_block_safe.ml



On Wed, Mar 23, 2016 at 3:31 PM, Hannes Mehnert <notifications@github.com>
wrote:

> I'm actually not sure about invalid input, and which module should
> validate that? Certainly, crashes are not nice, but shouldn't there (isn't
> there) a library which consumes a BLOCK device and presents a BLOCK device,
> just inserting bounds checks?
>
> In the end, a negative read (or other invalid reads/writes/...) should
> also be prevented from mirage-fs-xen, mirage-block-ramdisk,
> mirage-block-ccm, ... ? For me, a BLOCK in itself should not perform any
> bounds checks, but pretend that the input is already validated. (Of course,
> this can also be another thread of discussion, and a separate PR. Overall
> this PR improves things!)
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/mirage/mirage-fs-unix/pull/18#issuecomment-200395673>
>



-- 
Dave Scott

**Comment by talex5** [comment]
So... can we use @djs55's wrapper for this now?

**Pull request merged**